File: TaskManager/src/main/java/com/backend/taskmanager/service/AchievementService.java
package com.backend.taskmanager.service;

import com.backend.taskmanager.entity.Achievement;
import com.backend.taskmanager.entity.UserAchievement;
import java.util.List;
import java.util.Optional;

public interface AchievementService {

    // Get all achievements
    List<Achievement> getAllAchievements();

    // Get achievement by ID
    Optional<Achievement> findById(Long id);

    // Get achievements earned by a user
    List<UserAchievement> getUserAchievements(Long userId);

    // Get achievements a user hasn't earned yet
    List<Achievement> getUnearnedAchievements(Long userId);

    // Check and award achievements for a user
    void checkAndAwardAchievements(Long userId);

    // Count how many achievements a user has
    long countUserAchievements(Long userId);

    // Check if user has earned a specific achievement
    boolean hasUserEarnedAchievement(Long userId, Long achievementId);

    // Get progress toward a specific achievement
    int getAchievementProgress(Long userId, Long achievementId);
}

File: frontend/src/Services/achievementService.js
import API from "./api";

export const getUserAchievements = (userId) =>
  API.get(`/achievements/user/${userId}`);

export const getAvailableAchievements = (userId) =>
  API.get(`/achievements/user/${userId}/available`);

export const checkAchievements = (userId) =>
  API.post(`/achievements/user/${userId}/check`);

export const getAllAchievements = () => API.get("/achievements");

export const createAchievement = (data) => API.post("/achievements", data);

export const getAchievementProgress = (userId, achievementId) => 
  API.get(`/achievements/user/${userId}/progress/${achievementId}`);

export const getLeaderboard = (limit = 20) =>
    API.get(`/users/leaderboard?limit=${limit}`);

export const checkAndAwardAchievements = (userId) =>
  API.post(`/achievements/user/${userId}/check`);


